x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/last_value.tsx

```diff
@@ -0,0 +1,310 @@
+    const hasDateFields = indexPattern && getDateFields(indexPattern).length;
+    if (!hasDateFields) {
+      return i18n.translate('xpack.lens.indexPattern.lastValue.disabled', {
+        defaultMessage: 'This function requires the presence of a date field in your index',
+      });
+    }
+  },
+  getErrorMessage(layer, columnId, indexPattern) {
+    const column = layer.columns[columnId] as LastValueIndexPatternColumn;
+    let errorMessages: string[] = [];
+    const invalidSourceFieldMessage = getInvalidFieldMessage(column, indexPattern);
+    const invalidSortFieldMessage = getInvalidSortFieldMessage(
+      column.params.sortField,
+      indexPattern
+    );
+    if (invalidSourceFieldMessage) {
+      errorMessages = [...invalidSourceFieldMessage];
+    }
+    if (invalidSortFieldMessage) {
+      errorMessages = [invalidSortFieldMessage];
+    }
+    return errorMessages.length ? errorMessages : undefined;
+  },
+  buildColumn({ field, previousColumn, indexPattern }) {
+    const sortField = isTimeFieldNameDateField(indexPattern)
+      ? indexPattern.timeFieldName
+      : indexPattern.fields.find((f) => f.type === 'date')?.name;
+
+    if (!sortField) {
+      throw new Error(
+        i18n.translate('xpack.lens.functions.lastValue.missingSortField', {
+          defaultMessage: 'This index pattern does not contain any date fields',
+        })
+      );
+    }
+
+    const sortOrder =
+      (previousColumn?.params &&
+        'sortOrder' in previousColumn.params &&
+        previousColumn?.params?.sortOrder) ||
+      'desc';
+
+    return {
+      label: ofName(field.displayName),
+      dataType: field.type as DataType,
+      operationType: 'last_value',
+      isBucketed: false,
+      scale: field.type === 'string' ? 'ordinal' : 'ratio',
+      sourceField: field.name,
+      params: {
+        sortOrder,
+        sortField,
+      },
+    };
+  },
+  toEsAggsConfig: (column, columnId) => ({
+    id: columnId,
+    enabled: true,
+    schema: 'metric',
+    type: 'top_hits',
+    params: {
+      field: column.sourceField,
+      aggregate: 'concat',
+      size: 1,
+      sortOrder: column.params.sortOrder,
+      sortField: column.params.sortField,
+    },
+  }),
+
+  isTransferable: (column, newIndexPattern) => {
+    const newField = newIndexPattern.getFieldByName(column.sourceField);
+    const newTimeField = newIndexPattern.getFieldByName(column.params.sortField);
+    return Boolean(
+      newField &&
+        newField.type === column.dataType &&
+        !newField.aggregationRestrictions &&
+        newTimeField?.type === 'date'
+    );
+  },
+
+  paramEditor: ({ state, setState, currentColumn, layerId }) => {
+    const currentIndexPattern = state.indexPatterns[state.layers[layerId].indexPatternId];
+    const dateFields = getDateFields(currentIndexPattern);
+    const sortOrderButtons = [
+      {
+        id: `lns-lastValue-ascending`,
+        label: 'Ascending',
```

Comment:
I know I discussed it at length with @MichaelMarcialis, but I feel strongly that users will be confused by the words "Ascending" and "Descending", and even the word "sort order" as applied to this whole UI. We are pretty confident that "first value" is an edge case, so I would like to reorganize the form elements that make up this param editor.

In my opinion, the ideal form design is:

* Field selector
* Rename "sort by date field" to "time field"
* Move the sort order into an "advanced" popover
    * If using a radio button, label it as "Value mode: Last / First"
    * If using an EuiSwitch, label it as "Show first value"

This mock shows what I mean, except that I didn't create an advanced popover:

<img width="312" alt="Screen Shot 2020-12-01 at 3 45 57 PM" src="https://user-images.githubusercontent.com/666475/100795321-df80f100-33ec-11eb-99b4-33cb324d4c85.png">

Reply:
Thanks for the comment, @wylieconlon. I don't believe we got official closure after our last conversation on the matter, so I'll add the concerns I discussed with you here as well for visibility (and so we can continue the conversation on tomorrow's dev sync).

> Rename "sort by date field" to "time field"

I fear that renaming "sort by date field" to "time field" (and thus removing the action/verb from the label) could potentially confuse users. Yes, datetime fields are being presented and selectable, but that leaves the user wondering what selecting a field will do to their visualization? I believe telling them in the label that this selection is tied to sorting will greatly help in removing such confusion.

> Move the sort order into an "advanced" popover

I don’t believe it is necessary to hide the sort order button group in advanced settings for a few reasons. First, it is only a single field (which may feel odd to quarantine by itself in advanced settings). The sort order selection is also directly related to the preceding "sort by date field" selector (in which separating the two feels odd and may cause user confusion). Finally, the the current advanced setting paradigm would require three clicks to interact with it (one to open the advanced settings popover menu, one to select the advanced setting to add to the flyout, one to then change the sort order field). For these reasons, I personally prefer to leave it out of the advanced settings.

> If using a radio button, label it as "Value mode: Last / First"

I have concerns over user confusion when a function that is called "last value" also has a setting within it to choose between first/last value for the sort order. For example, users may be left wondering why there is an option to select a "first value" sort order in a function meant to return the last value. I feel using ascending/descending could potentially avoid such confusion.

Reply:
I just watched our meeting recording, and we did discuss this. We came to a very clear conclusion that "first value" is an edge case, but I don't think we came to a conclusion about the terminology. I've taken out the relevant parts of the transcript:

Q: Which fields should be sortable, and can we remove the Ascending and Descending options, and call them First and Last?

No clear answer in the recording.

@timroes:

> I'm not sure there's any good reason to support it on numerical fields, especially because the wording gets tricky. In my opinion, it's not first / last then, it's the whatever / whatever depending on the field.

@mbondyra:

> We could kind of call it first result

Q: Do we want to have "Last value" only, or keep it as "First / last", or as separate functions?

@AlonaNadler said:

> That's kind of what I was thinking. Let's call it Last value, and if they have the ability to sort it differently, they can get to first value. Last value and first value requires you to understand that it's the first value and last value in every bucket, and it's like last value is a little bit more simple and less open for interpretation.

@timroes said:

> I also haven't seen any use case where a user wanted specifically the first one from each bucket

> I would also think about putting it in an advanced section, to be honest, because I think we are really talking about more than 80% of people will want to leave it on last value and never switch it to first.

And finally, @AlonaNadler:

> Another sub-goal we have here is to reduce the number of decisions that we have in this huge form... let's try to minimize the amount of decisions and rely on good defaults.

---

My interpretation of this is that we should either remove the First value option or make it take extra clicks to select First value, because we don't want it to be confusing.

Reply:
The option has been removed just like we discussed in the sync.